### DISCUSSION TITLE
Generated count query is inefficient

### DISCUSSION THREAD

**Discussion by njeirath** [mikro-orm/mikro-orm#3044] [Ideas]
**Describe the bug**
I'm not sure this is a bug or if it's intended behavior but figured I'd raise it and seek some clarification.

When I'm performing a `findAndCount` operation, the count SQL that is being generated is using `count(distinct(e0.id))` rather than simply `count(*)`. 

Running an analyze on the `count(distinct(e0.id))` query reveals that it's running a full table scan:
```
explain analyze select count(distinct("e0"."id")) as "count" from "event" as "e0" where "e0"."end_time" < now();
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=3744.42..3744.43 rows=1 width=8) (actual time=120.039..120.041 rows=1 loops=1)
   ->  Seq Scan on event e0  (cost=0.00..3653.90 rows=36205 width=37) (actual time=0.008..16.095 rows=36206 loops=1)
         Filter: (end_time < now())
         Rows Removed by Filter: 2729
 Planning Time: 0.087 ms
 Execution Time: 120.074 ms
```

Whereas the simpler `count(*)`:
```
explain analyze select count(*) as "count" from "event" as "e0" where "e0"."end_time" < now();
                                                                  QUERY PLAN                                                                  
----------------------------------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=2986.88..2986.89 rows=1 width=8) (actual time=13.659..13.660 rows=1 loops=1)
   ->  Index Only Scan using end_time_idx on event e0  (cost=0.42..2896.36 rows=36205 width=0) (actual time=0.048..11.524 rows=36206 loops=1)
         Index Cond: (end_time < now())
         Heap Fetches: 5201
 Planning Time: 0.092 ms
 Execution Time: 13.689 ms
(6 rows)
```

The count(*) has almost a full order of magnitude better performance so I'm just wondering if there's a reason the count query is being generated the way it is.

**To Reproduce**
I'm specifying my entity and schemas separately, something similar to:

Entity:
```
export class Event {
    id: string
    endTime: Date
    ...
}
```

Schema:
```
export const schema = new EntitySchema<Event, Observable>({
    path: __filename,
    name: 'Event',
    class: Event,
    tableName: 'event',
    properties: {
        id: {type: 'string', primary: true},
        endTime: {type: 'datetime', nullable: false},
        ...
    }
})
```

And a simplified query `const [events, count] = await em.findAndCount(Event, {endTime: {$lt: new Date()}})

**Expected behavior**
I'd expect the count query to be using `count(*)` rather than counting by distinct id, especially since that ID is specified as the only PK on the table (unless of course I'm missing something here).

**Versions**

| Dependency | Version |
| - | - |
| node | 12 (lambda) |
| typescript | 3.8.3 |
| mikro-orm | 4.5.10 |
| your-driver | postgresql |

**Comment by B4nan** [comment]
Weird that it would be so slow. Note that `count(*)` is not the same as `count(distinct("e0"."id"))`. If we join to-many relations, there will be more rows that the actual count of root entities, so we need the disctinct clause. But if it would be really such overhead, we could detect whether its needed and use it only in case of the to-many joins.

**Comment by njeirath** [comment]
> Weird that it would be so slow. Note that count(*) is not the same as count(distinct("e0"."id")). If we join to-many relations, there will be more rows that the actual count of root entities, so we need the disctinct clause

Great point, I hadn't considered the to-many join scenario since my use case was only operating on a single table.

> But if it would be really such overhead, we could detect whether its needed and use it only in case of the to-many joins.

I did come across [this](https://stackoverflow.com/a/5203827/16920251) post that hints at why the table scan is occurring:

> If the SELECT returns more than approximately 5-10% of all rows in the table, a sequential scan is much faster than an index scan.

Our use case does in fact fall into this scenario, the query above is filtering out about 7% of the rows. However the overall # of rows is only around 36k, it appears what is really taking a lot of the time is the use of the distinct. Running an analyze on the same query but with the distinct removed gives me:

```
explain analyze select count("e0"."id") as "count" from "event" as "e0" where "e0"."end_time" < now();               
                                                     QUERY PLAN                                                       
---------------------------------------------------------------------------------------------------------------------
 Aggregate  (cost=3744.42..3744.43 rows=1 width=8) (actual time=14.049..14.050 rows=1 loops=1)
   ->  Seq Scan on event e0  (cost=0.00..3653.90 rows=36205 width=37) (actual time=0.009..12.204 rows=36206 loops=1)
         Filter: (end_time < now())
         Rows Removed by Filter: 2729
 Planning Time: 0.078 ms
 Execution Time: 14.077 ms
(6 rows)
```

This is still a slight bit slower than the `count(*)` but I'm assuming that's because it's still performing a table scan vs using the index in the `count(*)` version. I'm guessing as the table gets larger that the `count("e0"."id")` would have a larger difference compared to `count(*)`.

**Reply by B4nan** [reply]
This should do it:

https://github.com/mikro-orm/mikro-orm/commit/eebe34d8a11725c35b9e857f1ff4a1967cc6c1f8
